specification/flag-evaluation/flag-evaluation.md

```diff
@@ -0,0 +1,195 @@
+
+##### Requirement 1.2
+
+> The `API` **MUST** provide a function to set the `provider`, which accepts an API-conformant `provider` implementation.
+
+```
+// example provider mutator
+OpenFeature.setProvider(new MyProvider());
+```
+
+see [provider](../provider//providers.md)
+
+##### Requirement 1.3
+
+> The `API` **MUST** provide a function for `hook` addition, which accepts an array of API-conformant `hooks`. This adds the specified hooks to any hooks already attached.
+
+```
+// example hook attachment
+OpenFeature.addHooks([new MyHook()]);
+```
+
+see [hooks](./hooks.md)
+
+##### Requirement 1.4
+
+> The API **MUST** provide a function for retrieving the `provider` implementation.
+
+```
+// example provider accessor
+OpenFeature.getProvider();
+```
+
+see [provider](../provider/provider.md)
+
+##### Requirement 1.5
+
+> The `API` **MUST** provide a function for creating a `client`, with an options object containing an optional field for the client `name`.
+
+```
+// example client creation and retrieval
+OpenFeature.getClient({
+  name: 'my-openfeature-client'
+});
+```
+
+The name is a logical identifier for the client.
+
+##### Requirement 1.6
+
+> The client **MUST** provide a method for `hook` addition, which accepts an array of API-conformant `hooks`. This adds the specified hooks to any hooks already attached.
+
+```
+// example hook attachment
+client.addHooks([new MyHook()]);
+```
+
+see [hooks](./hooks.md)
+
+#### Flag Evaluation
+
+##### Requirement 1.7
+
+> The `client` **MUST** provide methods for typed flag value evaluation, including boolean, numeric, string, and map (object), with parameters `flag key` (string, required), `default value` (boolean | number | string | object, required), `evaluation context` (optional), and `evaluation options` (optional).
+
+```
+// example boolean flag evaluation
+boolean myBool =  client.getBooleanValue('bool-flag', false);
+
+
+// example overloaded string flag evaluation with optional params
+string myString = client.getStringValue('string-flag', 'N/A', evaluationContext, options);
+
+```
+
+see [evaluation context](../evaluation-context/evaluation-context.md)
+
+##### Requirement 1.8
+
+> The `client` **MUST** provide a method for flag activity evaluation, denoting if the flag is "enabled" or not, with parameters `flag key` (string, required), `default value` (boolean, required), `evaluation context` (optional), and `evaluation options` (optional).
+
+```
+// example flag activity evaluation
+boolean enabled = client.isEnabled('my-flag', false);
+```
+
+##### Requirement 1.9
+
+> The `client` method for flag activity evaluation **MAY** be semantically different from the boolean flag evaluation method.
+
+Some providers may draw a distinction between boolean flag value evaluation, and flag activity. Put another way, a flag may be "disabled", but simultaneously have an associated auxillary boolean payload of `true`.
```

Comment:
Is there additional information around this? I'm having some trouble parsing the use case that isn't handled internally in the provider.

Reply:
>This single API will tell you "Is this enabled in the UI" and also "is the flag evaluating to true"?

Not quite (unless I'm misunderstanding you)

The issue is  that some providers allow logically boolean flags (enabled/disabled) WITH separated associated boolean payloads (every flag has a bool and an optional number/string/boolean payload). Because some providers work that way, I think it forces us to have this kind of resolution - we need a method for the boolean flag, and a separate method to get it's associated boolean payload.

In providers that DONT have this concept, `isEnabled()` and `getBooleanValue()` do the same thing

Reply:
@InTheCloudDan 

>Is there additional information around this? I'm having some trouble parsing the use case that isn't handled internally in the provider.

See: https://github.com/open-feature/spec/pull/35/files#r855378994. It is really internally handled by the provider, but I wanted to call it out here because it's relevant to the expected behavior of the evaluation API.

Maybe this proposal isn't the best solution, but @dabeeeenster has brought the issue up a few times, and this is the best I can come up with. Open to more suggestions though.

Reply:
I'm also confused by this but it seems others are on top of clarifications

Reply:
@toddbaert Which providers do that? A flag evaluation that's "off" seems like it should return the default value and a reason for it.

Reply:
@justinabrahms  Flagsmith, particularly behaves this way. It can return false and a payload.

Reply:
Just to chime in here - we see the notion of a flag being `false` and consequently disabled or providing some form of convention in some way as something of an anti-pattern. There are use cases where flags are used in a manner that they travel from true > false for good reason - e.g. retiring a feature. 

We've designed Flagsmith in a manner that `false` does not impute anything in that regard - it's just something that is `false` and up to the developer to interpret that however they see fit.

Reply:
So, in Flagsmith, I can setup a single flag, let's call it `toddsFlag`. Like all flags in Flagsmith, it has a boolean and an optional `boolean | string | number` auxiliary value, depending on how you've configuration this it in this case. Let's consider the auxiliary flag is boolean for the rest of this discussion.

This same flag can used by the flagsmith SDK 2 ways:

`flagsmith.hasFeature('toddsFlag')` which returns a boolean corresponding to the big toggle switch on the Flagsmith UI. 

AND 

`flagsmith.getValue('toddsFlag')` which can return a boolean completely independent of the returned value of `hasFeature` above, true or false.

It was supporting this that I was attempting to do, but after this discussion, I'm wondering if it's not better to completely remove the `isEnabled()` method entirely, and let the Flagsmith provider determine how to handle resolving `getBooleanValue` in their case.

Reply:
Hmm thanks for the clarification. The wording here makes more sense with this in mind.

> It was supporting this that I was attempting to do, **but after this discussion, I'm wondering if it's not better to completely remove the `isEnabled()` method entirely**, and let the Flagsmith provider determine how to handle resolving `getBooleanValue` in their case.

I'm supportive of this tbh. Seems very provider-specific

Reply:
I'm in favor of removing `isEnabled()`. In my opinion, it seems very unlikely that an application author would expect `isEnabled()` and `getBooleanValue()` to behave differently. Removing it would eliminate that confusion.

Reply:
> flagsmith.getValue('toddsFlag') which can return a boolean completely independent of the returned value of hasFeature above, true or false. @dabeeeenster please correct me here, if I'm wrong.

Correct. 

This is something of a hangover from pure feature flags and a more complete flag and remote config solution. We wanted to give people a simple boolean state for a flag but then be able to enhance it with essentially a general String value, but then what if they store `True` in the string value, and how does that typing get supported through all the SDKs and their various type handling.

Hard to say what the best general solution is here. I think if the OpenFeature SDK can provide its best interpretation of the intention from the provider's interface but let developers have access to the base data is probably the optimal solution. 

I believe most vendors have a similar approach of booleans and additional config/multi-variate values but I cant (and don't want to !) speak for them.

Reply:
If this functionality is there due to historical reasons, I propose that we do not add this to the spec. I think the expected behavior for OpenFeature application developers would be that a disabled feature flag does not have any meaningful auxiliary value. In this case, it would simply return the default value provided during the evaluation. However, the provider pattern is flexible enough to allow each vendor to decide how best to implement the interface. In this case, the `getBooleanValue()` could get the value from the state of the flag or from the `getValue()` method.

Reply:
For now, I've removed the `isEnabled()` and `isEnabledDetails()` methods. We can always open up a new issue for this if we think it's something that's absolutely necessary, but I agree that the `provider` interface (spec to come) gives enough flexibility to address this in numerous ways.